src/mpid/ch4/netmod/ofi/ofi_win.c

```diff
@@ -189,56 +192,73 @@ static int win_allgather(MPIR_Win * win, void *base, int disp_unit)
-                                 0ULL,  /* In:  offset(not used)    */
-                                 MPIDI_OFI_WIN(win).mr_key,     /* In:  requested key       */
-                                 0ULL,  /* In:  flags               */
-                                 &MPIDI_OFI_WIN(win).mr,        /* Out: memregion object    */
-                                 NULL), mr_reg);        /* In:  context             */
+    /* fi_mr_reg may fail for legitimate reason, for example, with dynamic window, we
+     * are trying to register MPI_BOTTOM, some providers can legitimately reject.
+     *
+     * When fi_mr_reg fails, we should ensure always fallback to active message.
+     */
+
+    if (win->size > 0) {
+        int rc = fi_mr_reg(MPIDI_OFI_global.ctx[0].domain, base, win->size,
+                           FI_REMOTE_READ | FI_REMOTE_WRITE, 0ULL /* offset */ ,
+                           MPIDI_OFI_WIN(win).mr_key, 0ULL /* flags */ ,
+                           &MPIDI_OFI_WIN(win).mr, NULL /* context */);
+        if (rc < 0) {
+            /* Failed for whatever reason. Set mr_key to 0, which will enforce the fallback */
+            MPIDI_OFI_WIN(win).mr_key = MPIDI_OFI_INVALID_MR_KEY;
+        } else if (MPIDI_OFI_ENABLE_MR_PROV_KEY) {
+            /* retrieve provider key */
+            MPIDI_OFI_WIN(win).mr_key = fi_mr_key(MPIDI_OFI_WIN(win).mr);
+        }
+    } else {
+        /* note: this is "valid", we just won't use it */
+        MPIDI_OFI_WIN(win).mr_key = 0;
     }
+
     MPIDI_OFI_WIN(win).winfo = MPL_malloc(sizeof(*winfo) * comm_ptr->local_size, MPL_MEM_RMA);
 
     winfo = MPIDI_OFI_WIN(win).winfo;
     winfo[comm_ptr->rank].disp_unit = disp_unit;
-
-    if ((MPIDI_OFI_ENABLE_MR_PROV_KEY || MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS) && MPIDI_OFI_WIN(win).mr) {
-        /* MR_BASIC */
-        MPIDI_OFI_WIN(win).mr_key = fi_mr_key(MPIDI_OFI_WIN(win).mr);
-        winfo[comm_ptr->rank].mr_key = MPIDI_OFI_WIN(win).mr_key;
-        winfo[comm_ptr->rank].base = (uintptr_t) base;
-    }
+    winfo[comm_ptr->rank].mr_key = MPIDI_OFI_WIN(win).mr_key;
+    winfo[comm_ptr->rank].base = (uintptr_t) base;
 
     mpi_errno = MPIR_Allgather(MPI_IN_PLACE, 0,
                                MPI_DATATYPE_NULL,
                                winfo, sizeof(*winfo), MPI_BYTE, comm_ptr, &errflag);
     MPIR_ERR_CHECK(mpi_errno);
 
-    if (!MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS) {
-        first = winfo[0].disp_unit;
-        same_disp = 1;
-        for (i = 1; i < comm_ptr->local_size; i++) {
-            if (winfo[i].disp_unit != first) {
-                same_disp = 0;
-                break;
-            }
+    /* check we all got mr_key, or force fallback by setting mr to 0 */
+    for (i = 0; i < comm_ptr->local_size; i++) {
+        if (winfo[i].mr_key == MPIDI_OFI_INVALID_MR_KEY) {
+            MPIDI_OFI_WIN(win).mr = 0;
+            MPIDI_OFI_WIN(win).mr_key = MPIDI_OFI_INVALID_MR_KEY;
+            break;
         }
-        if (same_disp) {
```

Comment:
Be careful that, the code here is discarding a local MR based on remote mr_key. We need call `fi_close` to release the local MR.

Reply:
Good catch!